File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1:
            return head

        # sets first.next as last.next, and others next immediately preceding
        # sets pre_first.next or head to point to the new first 
        def reverse(first, last, pre_first):
            nonlocal head
            # take care of link to new first node
            if pre_first:
                pre_first.next = last
            else:
                head = last
        
            prev, curr = first, first.next
            first.next = last.next  # take care of link to remaining nodes
            for _ in range(k-1):
                curr.next, prev, curr = prev, curr, curr.next

        count = 1
        curr = first = head
        pre_first = None
        while curr:
            if count == k:
                # we have reached the kth node, now it's time to reverse and reset
                reverse(first, curr, pre_first)
                pre_first = first
                curr, first = first.next, first.next
                count = 1
            else:
                curr = curr.next
                count += 1
        
        return head
        

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head==NULL)
            return NULL;
        int temp=1;
        ListNode* tempNode=head;
        while(tempNode->next!= nullptr){
            tempNode=tempNode->next;
            temp++;
            if(temp>=k){
                break;
            }
        }
        if(temp<k){
            return head;
        }
        ListNode* next=NULL;
        ListNode* curr=head;
        ListNode* prev=NULL;
        int count=0;
        while(curr!=NULL&&count++<k){
            next=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next;  
        }
        head->next=reverseKGroup(curr,k);
        return prev;
    }
};
/*
class Solution {
 public:
  ListNode* reverseKGroup(ListNode* head, int k) {
    if (head == nullptr)
      return nullptr;

    ListNode* tail = head;

    for (int i = 0; i < k; ++i) {
      if (tail == nullptr)  // Less than k nodes, do nothing
        return head;
      tail = tail->next;
    }

    ListNode* newHead = reverse(head, tail);
    head->next = reverseKGroup(tail, k);
    return newHead;
  }

 private:
  // Reverses [head, tail)
  ListNode* reverse(ListNode* head, ListNode* tail) {
    ListNode* prev = nullptr;
    ListNode* curr = head;

    while (curr != tail) {
      ListNode* next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }

    return prev;
  }
};*/